### ISSUE TITLE
Insults terminal corruption

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8803]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)| @markrwilliams reported|
|-|-|
|Trac ID|trac#8803|
|Type|defect|
|Created|2016-09-03 00:35:19Z|
|Branch|https://github.com/twisted/twisted/tree/8803-insults-cursor-position|

As of 16.4.0, insults emits invalid control codes while positioning the cursor.

The fix is simple:

```
diff --git a/src/twisted/conch/insults/insults.py b/src/twisted/conch/insults/insults.py
index 3bcc182..e105b7d 100644
--- a/src/twisted/conch/insults/insults.py
+++ b/src/twisted/conch/insults/insults.py
@@ -721,7 +721,7 @@ class ServerProtocol(protocol.Protocol):
     def cursorPosition(self, column, line):
         self.write(b'\x1b[' +
                    intToBytes(line + 1)  +
-                   b'; '+
+                   b';' +
                    intToBytes(column + 1) +
                    b'H')

```

<details><summary>Searchable metadata</summary>

```
trac-id__8803 8803
type__defect defect
reporter__markrwilliams markrwilliams
priority__normal normal
milestone__None None
branch__8803_insults_cursor_position 8803-insults-cursor-position
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1472862919199874 1472862919199874
changetime__1473202048376885 1473202048376885
version__None None
owner__Glyph__glyph_____ Glyph <glyph@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_1"></a>|@markrwilliams commented|
|-|-|

Introduced here:

https://github.com/twisted/twisted/commit/b4b3b38259b159e1fe89944abf29137182a659b1#diff-3a08cb90858219a79d06564cea81740bL722

So this only affects 16.4.0.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

A `.1` for this spot-fix might be nice, since it is a clear regression.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_4"></a>|@markrwilliams commented|
|-|-|

https://github.com/twisted/twisted/pull/535

There are a bunch of tests with control code data lifted from http://www.ecma-international.org/publications/files/ECMA-ST/Ecma-048.pdf

Of course, not all the terminal control codes insults uses are defined in this standard.  But this should give us a place to start, and seems to have uncovered more Python 3 related issues.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

A totally beautiful fix; truly goes above and beyond test-coverage wise.  Will land.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @glyph<br>@glyph set status to `closed`|
|-|-|

In changeset f881ef231008d4bbf26b646b11d3e686ddff321a
```
#!CommitTicketReference repository="" revision="f881ef231008d4bbf26b646b11d3e686ddff321a"
Merge 8803-insults-cursor-position: pr 535

Author: markrwilliams

Reviewer: glyph

Fixes: #8803

twisted.conch.insults.insults.ServerProtocol no longer corrupts a client's display when attempting to set the cursor position, and its ECMA-48 terminal manipulation works on Python 3.
```
